src/main/webapp/login-page.html

```diff
@@ -33,5 +32,143 @@
+          // User is signed in.
+          if (user) {
+            var displayName = user.displayName
+            var email = user.email
+            var emailVerified = user.emailVerified
+            var photoURL = user.photoURL
+            var isAnonymous = user.isAnonymous
+            var uid = user.uid
+            var providerData = user.providerData
+
+            document.getElementById("login").textContent = "Log out"
+            document.getElementById("login-status").textContent = "Signed in."
+          } else {
+            // User signed out.
+            document.getElementById("login").textContent = "Log in"
+            document.getElementById("login-status").textContent = "Signed out."
+          }
+          document.getElementById("login").disabled = false
+        })
+      }
+
+      /**
+       * If user is currently logged in, log them out; if logged out, log them in.
+       */
+      function toggleLogin() {
+        // If the user is signed in, sign them out.
+        if (firebase.auth().currentUser) {
+          firebase.auth().signOut()
+        } else {
+          // Sign in.
+          var email = document.getElementById("email").value
+          var password = document.getElementById("password").value
+
+          if (email == null || email.length < 2) {
+            alert("Please enter an email address.")
+            return
+          }
+          if (password == null || password.length < 2) {
+            alert("Please enter a password.")
+            return
+          }
+
+          firebase
+            .auth()
+            .signInWithEmailAndPassword(email, password)
+            .catch(function onFailure(error) {
+              var errorCode = error.code
+              var errorMessage = error.message
+
+              if (errorCode === "auth/wrong-passcode") {
+                alert("Wrong password.")
+              } else {
+                alert(errorMessage)
+              }
+            })
+        }
+      }
+
+      /**
+       * Attempt to create a new user account using the email/password HTML fields.
+       */
+      function onSignup() {
+        var email = document.getElementById("email").value
+        var password = document.getElementById("password").value
+
+        if (email == null || email.length < 2) {
+          alert("Please enter a valid email address.")
+        }
+        if (password == null || password.length < 2) {
+          alert("Please enter a valid password.")
+        }
+
+        // If email/pass are of appropriate length, try to create a new account.
+        firebase
+          .auth()
+          .createUserWithEmailAndPassword(email, password)
+          .catch(function onFailure(error) {
+            var errorCode = error.code
+            var errorMessage = error.message
+
+            if (errorCode == "auth/weak-password") {
+              alert("The password is too weak. Passwords must be at least 6 characters long.")
```

Comment:
It is totally possible for firebase to redefine "weak-password" later (say, change it to 8 chars or required special chars) and this will be a really bad error message. You really use the provided error message here.